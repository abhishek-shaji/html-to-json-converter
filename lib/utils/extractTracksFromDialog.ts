import { TIME_BETWEEN_CHAR_MS } from '../constants/duration';
import { TRACK_TYPE_DIALOGUE } from '../constants/trackTypes';

export const extractTracksFromDialog = (
  dialog: Element,
  initialState: any,
  timestampStartAt: number
) => {
  const trackMap = { ...initialState };
  const children = [...dialog.children];

  for (const childElement of children) {
    const actionType = childElement.getAttribute('data-action-type');
    const target = childElement.getAttribute('data-action-target');
    const actionId = childElement.getAttribute('data-action-id');
    const timestamp =
      timestampStartAt + dialog.innerHTML.indexOf(childElement.outerHTML) * TIME_BETWEEN_CHAR_MS;

    if (!trackMap[actionType]) {
      trackMap[actionType] = [];
    }

    trackMap[actionType].push({ actionId, actionType, timestamp, target });

    childElement.remove();
  }

  const dialogueDuration = dialog.textContent.length * TIME_BETWEEN_CHAR_MS;

  if (!trackMap[TRACK_TYPE_DIALOGUE]) {
    trackMap[TRACK_TYPE_DIALOGUE] = [];
  }

  trackMap[TRACK_TYPE_DIALOGUE].push({
    dialog: dialog.textContent,
    duration: dialogueDuration,
    timestamp: timestampStartAt,
  });

  return { trackMap, dialogueDuration };
};
