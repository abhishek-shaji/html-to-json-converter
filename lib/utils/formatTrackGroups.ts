import { OutputType, TrackType } from '../types/output';

const formatTracks = (trackMap): TrackType[] =>
  Object.keys(trackMap).reduce((accum: TrackType[], type): TrackType[] => {
    return [
      ...accum,
      {
        type,
        actions: trackMap[type],
      },
    ];
  }, []);

export const formatTrackGroups = (trackGroupData): OutputType =>
  Object.keys(trackGroupData).reduce(
    (accum: OutputType, target): OutputType => {
      return {
        trackGroups: [
          ...accum.trackGroups,
          {
            target,
            tracks: formatTracks(trackGroupData[target]),
          },
        ],
      };
    },
    {
      trackGroups: [],
    }
  );
