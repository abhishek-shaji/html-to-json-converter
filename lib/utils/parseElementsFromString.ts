import { JSDOM } from 'jsdom';
import { EmptyFileException } from '../exceptions/EmptyFileException';

export const parseElementsFromString = (html: string): NodeListOf<Element> => {
  if (!html.trim()) {
    throw new EmptyFileException();
  }

  const dom = new JSDOM(html);
  const { document } = dom.window;

  return document.querySelectorAll('div.character, div.dialog');
};
