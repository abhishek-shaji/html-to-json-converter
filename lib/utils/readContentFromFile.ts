import { readFileSync } from 'fs';
import { FileReadFailureException } from '../exceptions/FileReadFailureException';

export const readContentFromFile = (path: string): string => {
  try {
    return readFileSync(path, 'utf8');
  } catch (err) {
    console.error(err);

    throw new FileReadFailureException();
  }
};
