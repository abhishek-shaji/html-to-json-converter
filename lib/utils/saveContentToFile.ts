import { writeFileSync, existsSync, mkdirSync } from 'fs';
import * as path from 'path';

const removeFilePart = (dirname) => path.dirname(dirname);

export const saveContentToFile = (file: string, data: string) => {
  const dirname = removeFilePart(file);

  if (!existsSync(dirname)) {
    mkdirSync(dirname, { recursive: true });
  }

  writeFileSync(file, data);
};
