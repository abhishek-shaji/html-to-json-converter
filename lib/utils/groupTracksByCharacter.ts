import { CLASS_NAME_DIALOG } from '../constants/classNames';
import { extractTracksFromDialog } from './extractTracksFromDialog';

export const groupTracksByCharacter = (elements: NodeListOf<Element>) => {
  const targets = {};
  let offset = 0;

  for (const [elementIndex, element] of elements.entries()) {
    const isDialog = element.classList.contains(CLASS_NAME_DIALOG);

    if (isDialog) {
      const character = elements[elementIndex - 1].innerHTML;

      if (!targets[character]) {
        targets[character] = {};
      }

      const { trackMap, dialogueDuration } = extractTracksFromDialog(
        element,
        targets[character],
        offset
      );

      offset = offset + dialogueDuration;

      targets[character] = trackMap;
    }
  }

  return targets;
};
