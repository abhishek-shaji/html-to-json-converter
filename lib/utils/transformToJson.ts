import { groupTracksByCharacter } from './groupTracksByCharacter';
import { formatTrackGroups } from './formatTrackGroups';

export const transformToJson = (elements: NodeListOf<Element>) => {
  const targets = groupTracksByCharacter(elements);
  const formattedData = formatTrackGroups(targets);

  return JSON.stringify(formattedData, null, 2);
};
