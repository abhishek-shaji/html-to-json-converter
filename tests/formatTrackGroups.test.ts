import { formatTrackGroups } from '../lib/utils/formatTrackGroups';

const mockData = {
  CH_medusa: {
    EM: [
      { actionId: 'EM_Laughing', actionType: 'EM', timestamp: 700, target: 'CH_genius' },
      { actionId: 'EM_Laughing', actionType: 'EM', timestamp: 700, target: 'CH_lucy' },
    ],
    DI: [
      {
        dialog: 'Every day I asked the cards when my baby was coming back. Isn’t that right, Lysi?',
        duration: 5670,
        timestamp: 0,
      },
    ],
  },
  CH_lucy: { DI: [{ dialog: 'Yes, Mama.', duration: 700, timestamp: 5670 }] },
  CH_genius: { DI: [{ dialog: 'Can’t breathe...', duration: 1120, timestamp: 6370 }] },
};

describe('formatTrackGroups', () => {
  it('should properly transform track group data to the desired output format', () => {
    const output = formatTrackGroups(mockData);

    expect(output).toMatchSnapshot();
  });
});
