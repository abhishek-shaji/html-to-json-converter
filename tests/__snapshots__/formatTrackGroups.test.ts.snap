// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatTrackGroups should properly transform track group data to the desired output format 1`] = `
{
  "trackGroups": [
    {
      "target": "CH_medusa",
      "tracks": [
        {
          "actions": [
            {
              "actionId": "EM_Laughing",
              "actionType": "EM",
              "target": "CH_genius",
              "timestamp": 700,
            },
            {
              "actionId": "EM_Laughing",
              "actionType": "EM",
              "target": "CH_lucy",
              "timestamp": 700,
            },
          ],
          "type": "EM",
        },
        {
          "actions": [
            {
              "dialog": "Every day I asked the cards when my baby was coming back. Isn’t that right, Lysi?",
              "duration": 5670,
              "timestamp": 0,
            },
          ],
          "type": "DI",
        },
      ],
    },
    {
      "target": "CH_lucy",
      "tracks": [
        {
          "actions": [
            {
              "dialog": "Yes, Mama.",
              "duration": 700,
              "timestamp": 5670,
            },
          ],
          "type": "DI",
        },
      ],
    },
    {
      "target": "CH_genius",
      "tracks": [
        {
          "actions": [
            {
              "dialog": "Can’t breathe...",
              "duration": 1120,
              "timestamp": 6370,
            },
          ],
          "type": "DI",
        },
      ],
    },
  ],
}
`;

exports[`formatTrackGroups should properly transform track group data to the output format 1`] = `
{
  "trackGroups": [
    {
      "target": "CH_medusa",
      "tracks": [
        {
          "actions": [
            {
              "actionId": "EM_Laughing",
              "actionType": "EM",
              "target": "CH_genius",
              "timestamp": 700,
            },
            {
              "actionId": "EM_Laughing",
              "actionType": "EM",
              "target": "CH_lucy",
              "timestamp": 700,
            },
          ],
          "type": "EM",
        },
        {
          "actions": [
            {
              "dialog": "Every day I asked the cards when my baby was coming back. Isn’t that right, Lysi?",
              "duration": 5670,
              "timestamp": 0,
            },
          ],
          "type": "DI",
        },
      ],
    },
    {
      "target": "CH_lucy",
      "tracks": [
        {
          "actions": [
            {
              "dialog": "Yes, Mama.",
              "duration": 700,
              "timestamp": 5670,
            },
          ],
          "type": "DI",
        },
      ],
    },
    {
      "target": "CH_genius",
      "tracks": [
        {
          "actions": [
            {
              "dialog": "Can’t breathe...",
              "duration": 1120,
              "timestamp": 6370,
            },
          ],
          "type": "DI",
        },
      ],
    },
  ],
}
`;
